### DISCUSSION TITLE
Full page image in digitized PDF

### DISCUSSION THREAD

**Discussion by chest3x** [pymupdf/PyMuPDF#994] [Looking for help] [answered]
Hi everyone, 
I'd like to ask, what is the optimal way to identify whether an image in a PDF page is of a size of a PDF page?
To explain further, I am having digitized PDF (with a textual layer), but there also is an image behind the entire page. I want to identify these documents and treat them separately.
Thanks in advance.

**Comment by JorjMcKie** [comment, marked as answer]
Scanned image pages usually have one (or more) images which cover the complete page - btw. not necessarily always being **_exactly_** the page size.
There is a ``Page`` method that computes the rectangle an image covers on the page, the "bbox" (boundary box). You can iterate through a page's images and check whether `page.rect in bbox`. Snippet:

```python
img_list = [img for img in page.get_images(True) if img[-1] == 0]  # only consider images directly called by page
for img in img_list:
    bbox = page.get_image_bbox(img)
    if page.rect in bbox:
        # fully covering image detected!
        break
```

Comments:
`get_images(True)` creates an extended image list, which also checks whether it is the page itself, that displays it, or some other object ("Form XObject").

**Reply by chest3x** [reply]
Thanks for your response. In case of the page image being split to multiple parts, I will calculate union of all images.

Also, one additional question. Is there a direct way to re-save a digitized PDF as a scanned PDF?
This basically means getting rid of the textual layer, only preserving the image, so each page would be an image only.

```
# My pdf_doc is a digitized PDF I want to turn into a scanned PDF (image only, no text present)

new_pdf = fitz.open()
for page in pdf_doc:
    page_image = page.get_pixmap()
    page_image.set_resolution(400, 400)
    ... # Can I save this Pixmap as a page, or is there any better approach
```

Thank You again.

**Reply by JorjMcKie** [reply]
You can of course do what you indicate with your snippet.
Except that your apparent intention (achieving a best possible image quality) will not be done with setting the resolution, but with makinbg a page pixmap  with a matrix that does this: ``pix = page.get_pixmap(matrix=fitz.Matrix(2,2))``. This will create an image with twice as many pixels in every direction, so 4-times the size.

Another, faster approach is possible with some of the other many PyMuPDF features: deleting the text directly and leaving everything else untouched:
```python
for page in doc:
    page.clean_contents()  # just making sure we have a single clean /Contents object
    xref = page.get_contents()[0]  # read its xref
    cont = bytearray(doc.xref_stream(xref))  # read it as a mutable bytearray
    while True:  # look for beginning of text objects
        p0 = cont.find(b"BT")
        if p0 < 0:
            break  # done
        p1 = cont.find(b"ET", p0)  # end of text object
        if p1 < 0:  # not found - should never happen!
            break
        cont[p0:p1] = b""  # remove text
    doc.update_stream(xref, cont)  # rewrite /Contents stream
    page.clean_contents()  # re-execute to get rid of any font references, too
```

**Reply by JorjMcKie** [reply]
Sorry about a typo in the previous code snippet: one code line must read ``cont[p0:p1+2] = b""  # remove text`` to remove the text object correctly. So the full snippet should be
```python
for page in doc:
    page.clean_contents()  # just making sure we have a single clean /Contents object
    xref = page.get_contents()[0]  # read its xref
    cont = bytearray(doc.xref_stream(xref))  # read it as a mutable bytearray
    while True:  # look for beginning of text objects
        p0 = cont.find(b"BT")
        if p0 < 0:
            break  # done
        p1 = cont.find(b"ET", p0)  # end of text object
        if p1 < 0:  # not found - should never happen!
            break
        cont[p0:p1+2] = b""  # remove text
    doc.update_stream(xref, cont)  # rewrite /Contents stream
    page.clean_contents()  # re-execute to get rid of any font references, too
```
